specs/GetSecurityDescriptorForPackageFamilyNames/GetSecurityDescriptorForPackageFamilyNames.md

```diff
@@ -0,0 +1,79 @@
+# Easing and streamlining named object sharing between packaged and Win32
+
+This feature is a flat C API that takes a list of Package Family Names (PFNs) and access masks, and returns a security descriptor
+
+# Background
+
+Named objects (mutexes, events, waitable timers) must be ACLed during creation in order to be shared with packaged applications.
+
+Currently, [our guidance](https://docs.microsoft.com/en-us/windows/uwp/communication/sharing-named-objects) for this scenario is to write a large amount of complex, cumbersome, and largely boilerplate code. ([See here for a taste](https://docs.microsoft.com/en-us/windows/win32/api/securityappcontainer/nf-securityappcontainer-getappcontainernamedobjectpath#examples)).
+
+If we are to meet developers where they are, we need to reduce the friction of Win32 and UWP interop, and make it as easy as possible to share named objects.
+
+# Description
+
+Add a flat C API that takes an array of PFNs and access masks, and produces the needed `SECURITY_ATTRIBUTES` object.
+
+# Examples
+
+```c++
+#include <wil/resource.h>
+
+wil::unique_event CreateShareableEvent(PCWSTR name)
+{
+    wil::unique_hlocal_ptr pSD;
+    PCWSTR packageNames[1] = { L"Contoso.Test.App_12345678" };
+    DWORD accessMasks[1] = { ACCESS_MODIFY_STATE };
+
+    THROW_IF_FAILED(GetSecurityDescriptorForPackageFamilyNames(
+        1, packageNames, accessMasks, &pSD);
+
+    SECURITY_ATTRIBUTES sa;
+    sa.nLength = sizeof(SECURITY_ATTRIBUTES);
+    sa.lpSecurityDescriptor = pSD.get();
+    sa.bInheritHandle = FALSE;
+
+    wil::unique_event result;
+    result.create(wil::EventOptions::None, name, &sa);
+    return result;
+}
+```
+
+# API Details
+
+```c
+STDAPI GetSecurityDescriptorForPackageFamilyNames(
```

Comment:
ForPackageFamilyNames or ForAppContainerNames?

AppContainers have a Name which is run through an algorithm to produce an AppContainerSID. PackageSID is just an AppContainerSID that's been computed using a PackageFamilyName as the AppContainer Name.

AppContainer Name is a string, w/o constraint I think (no restriction on content or length). PackageFamilyName has defined rules (length and content) so all valid PackageFamilyNames are a subset of all valid AppContainerNames.

Do you intend to error if pListOfPackageFamilyNames contains "Kittens", "Foo$Bar", "LPT1_1234567890abc" or other invalid PackageFamilyNames? If so then your implementation should
```
#include <appmodel.h>

STDAPI GetSecurityDescriptorForPackageFamilyNames(...) noexcept try
{
...
for (DWORD index=0; index < cCountOfPackageFamilyNames; ++index)
{
    auto packageFamilyName = pListOfPackageFamilyNames[index];
    THROW_IF_FAILED(VerifyPackageFamilyName(packageFamilyName));
}
...
}
CATCH_RETURN();
```

Functionally this API could work for AppContainerNames equally well as PackageFamilyNames (other than this parameter validation). Do you want to call it Get...ForAppContainerNames()? Or 2 functions, Get...ForAppContainerNames() doing the work and Get...ForPackageFamilyNames() that does the parameter validation then return Get...ForAppContainerNames()?
This function can operate on AppContainerNames as easily as PackageFamilyNames.

Reply:
Where do you see PackageFamilyName?
```
USERENVAPI HRESULT DeriveAppContainerSidFromAppContainerName(
  PCWSTR pszAppContainerName,
  PSID   *ppsidAppContainerSid
);
```

Yes, it accepts a PackageFamilyName because an AppContainerName is a string w/o constraint. All PackageFamilyNames are valid AppContainerNames...and so are a lot of values that aren't valid PackageFamilyNames.

GetSecurityDescriptorForPackageFamilyNames(..."Not.A$Package_Family@Name") should fail as that's not a valid package family name.

If the intent is to only support PackageFamilyName values then cool, please ensure the implementation validates its parameters.

If the intent is to support AppContainerNames then the function shouldn't validate parameters per `VerifyPackageFamilyName()` and it should be named *ForAppContainerNames().

Reply:
What are the scenarios where AppContainerName would be used over the more specific PackageFamilyName?

Reply:
If someone's using AppContainer without appmodel

AppContainer was invented in Win8 and is a core Windows technology, like CreateProcess() and ReadFile().
The appmodel uses AppContainer and is probably the most popular user, but you can use AppContainer w/o the appmodel. And several folks do.

Anyone who used LowIL would be interested in AppContainer, as a more secure alternative. Edge, Office and other non-packaged apps who historically used LowIL have jumped on AppContainer or been looking how to incorproate and/or switch to it. These folks by definition use AppContainerNames since they're not packaged, but they may find this API helpful just as packaged apps may.

Reply:
Ok. And how do you retrieve your AppContainer name in those scenarios?

Reply:
No retrieval. If you don't know your AppContainer name how'd you define it in the first place?

To create an AppContainer it needs to be defined to Windows via [CreateAppContainerProfile(PCWSTR pszAppContainerName, ...)](https://docs.microsoft.com/windows/win32/api/userenv/nf-userenv-createappcontainerprofile)

MSIX makes it convenient by always using PackageFamilyName, but if you subtract MSIX from the picture then you had to know your AppContainer name or you couldn't have create the AC profile in the first place.

Reply:
Ok. I was thinking also how you determine the other app's AppContainer name you're sharing with. PackageFamilyName has other ways like Get-AppxPackage and manifests to query the system for the values. If the apps are coordinating though or built by the same person, then it's just exchanged offline I guess.

I'm ok making this *ForAppContainerNames in this case. I think that's more encompassing and future proof. Then we just doc as part of the API what names are supported and how to get them.

Reply:
I'll update the API name.